sound/soc/sof/virtio-fe.c

```diff
@@ -438,7 +445,7 @@ struct snd_sof_dsp_ops snd_sof_vfe_ops = {
 
 	/* DAI drivers */
 	.drv		= virtio_dai,
-	.num_drv	= 1,
+	.num_drv	= 2,
```

Comment:
I don't get what SSP have to do in a file called virtio-fe. What sort of design are you implementing here? This is not mediation if you let each FE get access to a peripheral.

Reply:
virtio-fe is the uos virtual device driver. Comparing the SOS vBE driver, it doesn't know anything about the real HW connection. vFE will not operate directly on the HW, it will mediate with vBE with virtqueue, and vBE will operate on the real HW.

So, the design is, vFE1 <---> vBE,  vFE2 <----> vBE;  vBE <----> HW. With this mediator, every UOS can playback/capture. Otherwise, if vFE1 operates directly on the HW, other vFEs can't work.

Reply:
again no. vBE <--> PCM device, not hardware.

Reply:
I skip the PCM step here. Actually it should be:
vBE ---> find the PCM device related dai link ---> HW setting.

We didn't let vBE operate the PCM device directly because vFE we are still using the SOF architecture. So we can benefit from it and directly work with SOF to improve the performance.

vBE directly operates on PCM deivce is a VBS-U solution. The performance is worse than VBS-K. This is why we didn't operate on PCM device directly.